ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs

```diff
@@ -0,0 +1,662 @@
+  -- - Update the epoch nonce to the combination of the candidate nonce and the
+  --   nonce derived from the last block of the previous epoch.
+  -- - Update the "last block of previous epoch" nonce to the nonce derived from
+  --   the last applied block.
+  tickChainDepState
+    PraosConfig {praosEpochInfo}
+    (TickedPraosLedgerView lv)
+    slot
+    st =
+      TickedPraosState
+        { tickedPraosStateChainDepState = st',
+          tickedPraosStateLedgerView = TickedPraosLedgerView lv
+        }
+      where
+        newEpoch =
+          isNewEpoch
+            (History.toPureEpochInfo praosEpochInfo)
+            (praosStateLastSlot st)
+            slot
+        st' =
+          if newEpoch
+            then
+              st
+                { praosStateEpochNonce =
+                    praosStateCandidateNonce st
+                      ⭒ praosStateLastEpochBlockNonce st,
+                  praosStateLastEpochBlockNonce = praosStateLabNonce st
+                }
+            else st
+
+  -- Validate and update the chain dependent state as a result of processing a
+  -- new header.
+  --
+  -- This consists of:
+  -- - Validate the VRF checks
+  -- - Validate the KES checks
+  -- - Call 'reupdateChainDepState'
+  --
+  updateChainDepState
+    cfg@( PraosConfig
+            PraosParams {praosLeaderF}
+            _
+          )
+    b
+    slot
+    tcs = do
+      -- First, we check the KES signature, which validates that the issuer is
+      -- in fact who they say they are.
+      validateKESSignature cfg lv (praosStateOCertCounters cs) b
+      -- Then we examing the VRF proof, which confirms that they have the
+      -- right to issue in this slot.
+      validateVRFSignature (praosStateEpochNonce cs) lv praosLeaderF b
+      -- Finally, we apply the changes from this header to the chain state.
+      pure $ reupdateChainDepState cfg b slot tcs
+      where
+        lv = getTickedPraosLedgerView (tickedPraosStateLedgerView tcs)
+        cs = tickedPraosStateChainDepState tcs
+
+  -- Re-update the chain dependent state as a result of processing a header.
+  --
+  -- This consists of:
+  -- - Update the last applied block hash.
+  -- - Update the evolving and (potentially) candidate nonces based on the
+  --   position in the epoch.
+  -- - Update the operational certificate counter.
+  reupdateChainDepState
+    _cfg@( PraosConfig
+             PraosParams {praosSecurityParam, praosLeaderF}
+             ei
+           )
+    b
+    slot
+    tcs =
+      cs
+        { praosStateLastSlot = NotOrigin slot,
+          praosStateLabNonce = prevHashToNonce (Views.hvPrevHash b),
+          praosStateEvolvingNonce = newEvolvingNonce,
+          praosStateCandidateNonce =
+            if slot +* Duration stabilityWindow < firstSlotNextEpoch
+              then newEvolvingNonce
+              else praosStateCandidateNonce cs,
+          praosStateOCertCounters =
+            Map.insert hk n $ praosStateOCertCounters cs
```

Comment:
Micro-optimisation opportunity (perhaps as a TODO): 99% of the time this update will be an identity. If we do a lookup first, we will save a few allocations.